lib/widgets/profile.dart

```diff
@@ -0,0 +1,259 @@
+  }
+}
+
+class _ProfileErrorPage extends StatelessWidget {
+  const _ProfileErrorPage();
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(title: const Text('Error')),
+      body: const SingleChildScrollView(
+        child: Padding(
+          padding: EdgeInsets.fromLTRB(16, 32, 16, 16),
+          child: Row(
+            mainAxisAlignment: MainAxisAlignment.center,
+            children: [
+              Icon(Icons.error),
+              Text('Could not show user profile.'),
+            ]))));
+  }
+}
+
+String roleToLabel(UserRole role) {
+  return switch (role) {
+    UserRole.owner => 'Owner',
+    UserRole.administrator => 'Administrator',
+    UserRole.moderator => 'Moderator',
+    UserRole.member => 'Member',
+    UserRole.guest => 'Guest',
+    UserRole.unknown => 'Unknown',
+  };
+}
+
+class _ProfileDataTable extends StatelessWidget {
+  const _ProfileDataTable({required this.profileData});
+
+  final Map<int, ProfileFieldUserData>? profileData;
+
+  static T? _safeDecode<T>(T Function(Map<String, dynamic>) fromJson, String data) {
+    try {
+      return fromJson(jsonDecode(data));
+    } on FormatException {
+      return null;
+    } on TypeError {
+      return null;
+    }
+  }
+
+  Widget? _buildCustomProfileFieldValue(BuildContext context, String value, CustomProfileField realmField) {
+    final store = PerAccountStoreWidget.of(context);
+
+    switch (realmField.type) {
+      case CustomProfileFieldType.link:
+        return _LinkWidget(url: value, text: value);
+
+      case CustomProfileFieldType.choice:
+        final choiceFieldData = _safeDecode(CustomProfileFieldChoiceDataItem.parseFieldDataChoices, realmField.fieldData);
+        if (choiceFieldData == null) return null;
+        final choiceItem = choiceFieldData[value];
+        return (choiceItem == null) ? null : _TextWidget(text: choiceItem.text);
+
+      case CustomProfileFieldType.externalAccount:
+        final externalAccountFieldData = _safeDecode(CustomProfileFieldExternalAccountData.fromJson, realmField.fieldData);
+        if (externalAccountFieldData == null) return null;
+        final urlPattern = externalAccountFieldData.urlPattern ??
+          store.realmDefaultExternalAccounts[externalAccountFieldData.subtype]?.urlPattern;
+        if (urlPattern == null) return null;
+        final url = urlPattern.replaceFirst('%(username)s', value);
+        return _LinkWidget(url: url, text: value);
+
+      case CustomProfileFieldType.user:
+        // TODO(server): This is completely undocumented.  The key to
+        //   reverse-engineering it was:
+        //   https://github.com/zulip/zulip/blob/18230fcd9/static/js/settings_account.js#L247
+        final List<int> userIds;
+        try {
+          userIds = (jsonDecode(value) as List<dynamic>).map((e) => e as int).toList();
+        } on FormatException {
+          return null;
+        } on TypeError {
+          return null;
+        }
```

Comment:
Ah, I guess this also requires the signature of `_safeDecode` to get relaxed a bit, because this one wants to process a List whereas the others want Map. I think this will do the job:
```dart
  static T? _safeDecode<T, U>(T Function(U) fromJson, String data) {
```

And then here:

```suggestion
        final userIds = _safeDecode((List<dynamic> json) {
          return json.map((e) => e as int).toList();
        }, value);
        if (userIds == null) return null;
```